.travis.yml

```diff
@@ -17,7 +17,7 @@ notifications:
     - thomas_d_j@yahoo.com.au
 
 script:
-    - scons VERBOSE=1
+    - scons VERBOSE=1 DEBUG=1 CCFLAGS=-g0
```

Comment:
Extra Testing
-----------

I agree that it should be as close to the user's setup as possible; however, I always thought of a debug build as being:

* `the user's setup` + `a lot of extra runtime checks` - `optimizations`

For instance, without a debug build, the `g_assert()` calls are removed, which removes testing for programming errors, thereby potentially letting bugs through.

I suppose one could worry about bugs being introduced by the optimizations, and thus one might desire to test an optimized program. I have 2 points on that:

* One solution is to allow for an optimized build with assertions still enabled (that is, decouple optimization and assertions).
* Another solution is to trust that the compiler makers have tested their optimizations well enough, so that one may assume they don't introduce bugs (i.e., just not worry about it).

Personally, I lean towards trusting the compilers, but I can also see the virtue in decoupling optimization and assertions.

Efficiency of Testing
---------------------

However, optimizations are not necessarily a good thing here, as they must be performed every single time the tests are run.

<details>
<summary>To evaluate this, here are some bash functions.</summary>

    #!/bin/bash
    
    msg() { printf "$@" >&3; }
    
    iteration() { msg '(%s of 2) %s\n' "$1" "$2"; }
    
    test()
    (
        set -e
        exec 4>"$1" 3>&1 2>/dev/null 1>&2; shift
        command="scons ${1:+"$1 "}test"
        git clean -fdx
        iteration 1 "$command"
        (
            time \
            {
                eval "$command"
                git clean -fdx
                iteration 2 "$command"
                eval "$command"
            } 2>/dev/null
        ) 2>&4
    )
    
    separate()
    {
        (($#)) &&
            msg '\n%s\n' "=========== $1:" ||
            msg '%s\n' -----------
    }
    
    compare()
    (
        set -e
        exec 3>&1 2>/dev/null 1>&2
    
        name1=${1%%:*}
        if [[ -z $name1 ]]; then
            name1=Test\ 1
        fi
        args1=${1#*:}
    
        name2=${2%%:*}
        if [[ -z $name2 ]]; then
            name2=Test\ 2
        fi
        args2=${2#*:}
    
        separate 'Setup'
        msg "Checking out \"$head\"... "
        git clean -fdx
        git checkout "$head"
        msg 'finished\n'
    
        msg 'Warmup... '
        eval scons "$args1" test
        msg 'finished\n'
        separate 'Running Tests'
        test /tmp/test1 "$args1" 1>&3
        separate
        test /tmp/test2 "$args2" 1>&3
    
        separate "$name1"
        cat /tmp/test1 >&3
        separate "$name2"
        cat /tmp/test2 >&3
    )
</details><br>

The time it takes to build and test a non-debug build can be compared easily to the corresponding time that results from the proposed changes:

    $ head=bff796458872a1781199cd72eaef30f7f891218d
    $ compare Normal: Debug:'DEBUG=1 CCFLAGS=-g0'
    
    =========== Setup:
    Checking out "bff796458872a1781199cd72eaef30f7f891218d"... finished
    Warmup... finished
    
    =========== Running Tests:
    (1 of 2) scons test
    (2 of 2) scons test
    -----------
    (1 of 2) scons DEBUG=1 CCFLAGS=-g0 test
    (2 of 2) scons DEBUG=1 CCFLAGS=-g0 test
    
    =========== Normal:
    
    real    7m30.190s
    user    3m31.650s
    sys     0m46.404s
    
    =========== Debug:
    
    real    7m16.059s
    user    3m11.495s
    sys     0m46.223s

I ran that command twice, and came up with the same results:

* The proposed changes make the testing significantly faster (by about 10 seconds per cycle on my machine).

Because this program is undoubtedly I/O bound, and because the debug build performs many more run-time checks, I suspect that a great deal of the savings come from the compilation phase. A further test bears this out (the `#` comments out the `test` target, thereby running just a build):

    $ compare Normal:'#' Debug:'DEBUG=1 CCFLAGS=-g0 #'
    
    =========== Setup:
    Checking out "bff796458872a1781199cd72eaef30f7f891218d"... finished
    Warmup... finished
    
    =========== Running Tests:
    (1 of 2) scons # test
    (2 of 2) scons # test
    -----------
    (1 of 2) scons DEBUG=1 CCFLAGS=-g0 # test
    (2 of 2) scons DEBUG=1 CCFLAGS=-g0 # test
    
    =========== Normal:
    
    real    1m47.417s
    user    1m37.621s
    sys     0m6.775s
    
    =========== Debug:
    
    real    1m8.520s
    user    0m59.875s
    sys     0m6.445s

Wow! Look at that! Each debug build is around 19 seconds faster on my machine.

Conclusion
----------

Therefore, *currently*, the proposed changes maximize the usefulness of the testing cycle; the result is a wider search for bugs and yet a faster turnaround.

What do you think?

Reply:
> For instance, without a debug build, the g_assert() calls are removed, which removes testing for programming errors, thereby potentially letting bugs through.

You have a point there, I forgot about the new ``g_assert()`` relation with ``DEBUG=1``.

> Personally, I lean towards trusting the compilers, but I can also see the virtue in decoupling optimization and assertions.

I would lean to introducing to decoupling of assertions and debug symbols. You never know what compiler/versions of compilers people use to compile ``rmlint``. 

> Therefore, currently, the proposed changes maximize the usefulness of the testing cycle; the 
 result is a wider search for bugs and yet a faster turnaround.
>
> What do you think?

Well, nice... but: We're still talking about the setup on TravisCI here. A service that runs in background and is usually (rightly) totally forgotten until something breaks. It's really one of the last things I'd ever optimize. If it takes a few seconds more to send me a mail notification I wouldn't even notice it. I would not trade it against testing an optimized build.